File: chemsim/ops/heat_exchanger.py
"""Two-stream heat exchanger."""
from __future__ import annotations

import copy
from enum import Enum

from chemsim.core import Phase, Stream
from chemsim.ops.base import IUnitOp
from chemsim.thermo.flash import FlashCalculator


class HXSpec(Enum):
    DUTY = "DUTY"
    HOT_OUTLET_T = "HOT_OUTLET_T"


class HeatExchangerOp(IUnitOp):
    def __init__(self, spec: HXSpec,
                 fc_hot: FlashCalculator,
                 fc_cold: FlashCalculator):
        self._spec = spec
        self._fc_hot = fc_hot
        self._fc_cold = fc_cold
        self.hot_in = Stream()
        self.cold_in = Stream()
        self.hot_out = Stream()
        self.cold_out = Stream()
        self.Q_spec: float = 0.0      # W, used when spec=DUTY
        self.T_hot_out: float = 300.0  # K, used when spec=HOT_OUTLET_T
        self.duty: float = 0.0

    def inlet_ports(self):
        return ["hot_in", "cold_in"]

    def outlet_ports(self):
        return ["hot_out", "cold_out"]

    def set_inlet(self, port: str, stream: Stream) -> None:
        if port == "hot_in":
            self.hot_in = stream
        elif port == "cold_in":
            self.cold_in = stream
        else:
            raise ValueError(f"HeatExchangerOp: unknown inlet port '{port}'")

    def get_outlet(self, port: str) -> Stream:
        if port == "hot_out":
            return self.hot_out
        if port == "cold_out":
            return self.cold_out
        raise ValueError(f"HeatExchangerOp: unknown outlet port '{port}'")

    @staticmethod
    def _stream_H(fc: FlashCalculator, s: Stream) -> float:
        r = fc.flash_TP(s.T, s.P, s.z)
        return fc.total_enthalpy(r)

    def solve(self) -> None:
        if self.hot_in.total_flow <= 0.0:
            raise ValueError("HeatExchangerOp: hot_in.total_flow <= 0")
        if self.cold_in.total_flow <= 0.0:
            raise ValueError("HeatExchangerOp: cold_in.total_flow <= 0")

        H_hot_in  = self._stream_H(self._fc_hot,  self.hot_in)
        H_cold_in = self._stream_H(self._fc_cold, self.cold_in)

        if self._spec == HXSpec.DUTY:
            Q = self.Q_spec
        else:
            r_h = self._fc_hot.flash_TP(self.T_hot_out, self.hot_in.P, self.hot_in.z)
            H_hot_out_spec = self._fc_hot.total_enthalpy(r_h)
            Q = self.hot_in.total_flow * (H_hot_in - H_hot_out_spec)

        self.duty = Q

        H_hot_out  = H_hot_in  - Q / self.hot_in.total_flow
        H_cold_out = H_cold_in + Q / self.cold_in.total_flow

        self.hot_out  = self._flashPH_stream(self._fc_hot,  self.hot_in,  H_hot_out)
        self.cold_out = self._flashPH_stream(self._fc_cold, self.cold_in, H_cold_out)

    def _flashPH_stream(self, fc: FlashCalculator,
                        inlet: Stream, H_out: float) -> Stream:
        r = fc.flash_PH(inlet.P, H_out, inlet.z, inlet.T)
        out = copy.copy(inlet)
        out.T = r.T; out.vapor_fraction = r.beta
        out.x = r.x; out.y = r.y
        out.phase = (Phase.LIQUID if r.beta < 1e-10
                     else Phase.VAPOR if r.beta > 1.0 - 1e-10
                     else Phase.MIXED)
        out.H = fc.total_enthalpy(r)
        out.S = fc.total_entropy(r)
        return out

File: src/ops/HeatExchanger.cpp
#include "chemsim/ops/HeatExchanger.hpp"
#include <stdexcept>

namespace chemsim {

HeatExchanger::HeatExchanger(Spec spec,
                             const FlashCalculator& fc_hot,
                             const FlashCalculator& fc_cold)
    : spec_(spec), fc_hot_(fc_hot), fc_cold_(fc_cold) {}

double HeatExchanger::streamH(const FlashCalculator& fc, const Stream& s) {
    // Compute total specific enthalpy [J/mol] from stream conditions
    auto r = fc.flashTP(s.T, s.P, s.z);
    return fc.totalEnthalpy(r);
}

void HeatExchanger::solve() {
    if (hot_in.totalFlow  <= 0.0) throw std::invalid_argument("HeatExchanger: hot_in.totalFlow <= 0");
    if (cold_in.totalFlow <= 0.0) throw std::invalid_argument("HeatExchanger: cold_in.totalFlow <= 0");

    double H_hot_in  = streamH(fc_hot_,  hot_in);
    double H_cold_in = streamH(fc_cold_, cold_in);

    double Q;  // [W], heat transferred hot→cold
    if (spec_ == Spec::DUTY) {
        Q = Q_spec;
    } else {  // HOT_OUTLET_T
        // Flash hot side at specified outlet T to get H_hot_out
        auto r_h = fc_hot_.flashTP(T_hot_out, hot_in.P, hot_in.z);
        double H_hot_out_spec = fc_hot_.totalEnthalpy(r_h);
        Q = hot_in.totalFlow * (H_hot_in - H_hot_out_spec);
    }

    duty = Q;

    // Hot outlet
    double H_hot_out = H_hot_in - Q / hot_in.totalFlow;
    auto r_hot = fc_hot_.flashPH(hot_in.P, H_hot_out, hot_in.z, hot_in.T);
    hot_out           = hot_in;
    hot_out.T         = r_hot.T;
    hot_out.vaporFraction = r_hot.beta;
    hot_out.x         = r_hot.x;
    hot_out.y         = r_hot.y;
    hot_out.phase     = (r_hot.beta < 1e-10)  ? Phase::LIQUID
                      : (r_hot.beta > 1-1e-10) ? Phase::VAPOR
                                               : Phase::MIXED;
    hot_out.H = fc_hot_.totalEnthalpy(r_hot);
    hot_out.S = fc_hot_.totalEntropy (r_hot);

    // Cold outlet
    double H_cold_out = H_cold_in + Q / cold_in.totalFlow;
    auto r_cold = fc_cold_.flashPH(cold_in.P, H_cold_out, cold_in.z, cold_in.T);
    cold_out           = cold_in;
    cold_out.T         = r_cold.T;
    cold_out.vaporFraction = r_cold.beta;
    cold_out.x         = r_cold.x;
    cold_out.y         = r_cold.y;
    cold_out.phase     = (r_cold.beta < 1e-10)  ? Phase::LIQUID
                       : (r_cold.beta > 1-1e-10) ? Phase::VAPOR
                                                 : Phase::MIXED;
    cold_out.H = fc_cold_.totalEnthalpy(r_cold);
    cold_out.S = fc_cold_.totalEntropy (r_cold);
}

} // namespace chemsim
